File: testing-unit-cs-mocks/Testing.cs
// Copyright 2016-2020, Pulumi Corporation

using System.Collections.Generic;
using System.Collections.Immutable;
using System.Threading.Tasks;
using Pulumi;
using Pulumi.Testing;

namespace UnitTesting
{
    class Mocks : IMocks
    {
        public Task<(string? id, object state)> NewResourceAsync(MockResourceArgs args)
        {
            var outputs = ImmutableDictionary.CreateBuilder<string, object>();

            // Forward all input parameters as resource outputs, so that we could test them.
            if (args.Inputs != null)
            {
                outputs.AddRange(args.Inputs);
            }

            // Set the name to resource name if it's not set explicitly in inputs.
            if (args.Inputs?.ContainsKey("name") != true && args.Name != null)
            {
                outputs.Add("name", args.Name);
            }

            if (args.Type == "azure-native:storage:Blob")
            {
                // Assets can't directly go through the engine.
                // We don't need them in the test, so blank out the property for now.
                outputs.Remove("source");
            }

            // For a Storage Account...
            if (args.Type == "azure-native:storage:StorageAccount" && args.Name != null)
            {
                // ... set its web endpoint property.
                // Normally this would be calculated by Azure, so we have to mock it.
                var endpoints = new Dictionary<string, string>
                {
                    { "web", $"https://{args.Name}.web.core.windows.net" }
                };
                outputs.Add("primaryEndpoints", endpoints.ToImmutableDictionary());
            }

            // Default the resource ID to `{name}_id`.
            args.Id ??= args.Name != null ? $"{args.Name}_id" : "unknown_id";
            return Task.FromResult((args.Id, (object)outputs.ToImmutable()));
        }

        public Task<object> CallAsync(MockCallArgs args)
        {
            // We don't use this method in this particular test suite.
            // Default to returning whatever we got as input.
            return Task.FromResult((object)args.Args);
        }
    }

    public static class TestingExtensions
    {
        /// <summary>
        /// Extract the value from an output.
        /// </summary>
        public static Task<T> GetValueAsync<T>(this Output<T> output)
        {
            var tcs = new TaskCompletionSource<T>();
            output.Apply(v =>
            {
                tcs.SetResult(v);
                return v;
            });
            return tcs.Task;
        }
    }
}

File: testing-unit-fs-mocks/Testing.fs
// Copyright 2016-2020, Pulumi Corporation

namespace UnitTesting

open System
open System.Collections.Generic
open System.Collections.Immutable
open System.Threading.Tasks
open Pulumi.Testing

type Mocks() =
    interface IMocks with
        member this.NewResourceAsync(args: MockResourceArgs): Task<ValueTuple<string,obj>> =
            let inputKVs = args.Inputs |> Seq.map(fun kv -> kv.Key, kv.Value) |> Seq.toList
            let nameKVs = if args.Inputs.ContainsKey("name") then [] else [("name", args.Name :> obj)]
            let endpointKVs =
                if args.Type = "azure:storage/account:Account"
                then ["primaryWebEndpoint", sprintf "https://%s.web.core.windows.net" args.Name :> obj]
                else []
            
            let outputs =
                [inputKVs; nameKVs; endpointKVs]
                |> Seq.concat
                |> Seq.filter (fun (k, _) -> args.Type <> "azure:storage/blob:Blob" || k <> "source")
                |> Seq.map KeyValuePair
            let dict = outputs.ToImmutableDictionary() :> obj            

            // Default the resource ID to `{name}_id`.
            let id = if args.Id = null then sprintf "%s_id" args.Name else args.Id 
            Task.FromResult(struct (id, dict))

        member this.CallAsync(args: MockCallArgs): Task<obj> =
            // We don't use this method in this particular test suite.
            // Default to returning whatever we got as input.
            Task.FromResult(null :> obj)
